libsrc/c64/emd/c64-kerberos.s

```diff
@@ -0,0 +1,280 @@
+RAMC_PAGE_HI      = $DE3F                 ; Page register high
+
+.code
+
+; ------------------------------------------------------------------------
+; INSTALL routine. Is called after the driver is loaded into memory. If
+; possible, check if the hardware is present and determine the amount of
+; memory available.
+; Must return an EM_ERR_xx code in a/x.
+;
+
+INSTALL:
+	;; write $55 into first page
+	lda	#0
+	sta	RAMC_PAGE_LO
+	sta	RAMC_PAGE_HI
+	lda	#$55
+	sta	RAMC_WINDOW
+
+	;; write $AA into second page
+	lda	#1
+	sta	RAMC_PAGE_LO
+	sta	RAMC_PAGE_HI
+	lda	#$AA
+	sta	RAMC_WINDOW
+
+	;; check $55 in first page
+	lda	#0
+	sta	RAMC_PAGE_LO
+	sta	RAMC_PAGE_HI
+	lda	RAMC_WINDOW
+	cmp	#$55
+	bne	@notpresent
+
+	;; check $AA in first page
+	lda	#1
+	sta	RAMC_PAGE_LO
+	sta	RAMC_PAGE_HI
+	lda	RAMC_WINDOW
+	cmp	#$AA
+	bne	@notpresent
+
+	lda     #<EM_ERR_OK
+        ldx     #>EM_ERR_OK
+	rts
+
+@notpresent:
+        lda     #<EM_ERR_NO_DEVICE
+        ldx     #>EM_ERR_NO_DEVICE
+	; use rts from UNINSTALL below
+
+; ------------------------------------------------------------------------
+; UNINSTALL routine. Is called before the driver is removed from memory.
+; Can do cleanup or whatever. Must not return anything.
+;
+
+UNINSTALL:
+        rts
+
+; ------------------------------------------------------------------------
+; PAGECOUNT: Return the total number of available pages in a/x.
+;
+
+PAGECOUNT:
+        lda     #0
+        ldx     #2
+        rts
+
+; ------------------------------------------------------------------------
+; USE: Tell the driver that the window is now associated with a given page.
+; The Kerberos cartridge does not copy but actually map the window, so USE is
+; identical to MAP.
+
+USE     = MAP
+
+; ------------------------------------------------------------------------
+; MAP: Map the page in a/x into memory and return a pointer to the page in
+; a/x. The contents of the currently mapped page (if any) may be discarded
+; by the driver.
+;
+
+MAP:    sta     RAMC_PAGE_LO
+	txa
+	and	#1
+        sta     RAMC_PAGE_HI
+        lda     #<RAMC_WINDOW
+        ldx     #>RAMC_WINDOW
+
+; Use the RTS from COMMIT below to save a precious byte of storage
+
+; ------------------------------------------------------------------------
+; COMMIT: Commit changes in the memory window to extended storage.
+
+COMMIT: rts
+
+; ------------------------------------------------------------------------
+; COPYFROM: Copy from extended into linear memory. A pointer to a structure
+; describing the request is passed in a/x.
+; The function must not return anything.
+;
+
+COPYFROM:
+        jsr     setup
+
+; Setup is:
+;
+;   - ptr1 contains the struct pointer
+;   - ptr2 contains the linear memory buffer
+;   - ptr3 contains -(count-1)
+;   - tmp1 contains the low page register value
+;   - tmp2 contains the high page register value
+;   - X contains the page offset
+;   - Y contains zero
+
+        jmp     @L5
```

Comment:
That instruction can be
```asm
beq     @L5
```
because of the `ldy #$00` at the end of `setup`.

Reply:
done.